### PR TITLE
upgrade pinned builds to clang 10 & boost 1.72

### PR DESCRIPTION
**Pull request by spoonincode**
## Change Description

The state DB layout has changed in develop: any post-2.0 release will require users to migrate via a portable snapshot. Might as well take the opportunity to bump our pinned compiler environment.

Upgrade clang from 8.0 to 10.0. Upgrade Boost from 1.71.0 to 1.72.0. Non-optimized replay performance shows a marginal but measurable performance gain. It's nice to have the latest anyways, right?

Also took this opportunity to migrate our clang/llvm clone to the official GitHub monorepo.

This is being left as draft until clang 10.0 final is released, and documentation may need adjustment too.

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

**Comment by spoonincode** [comment]
I had @emorybarlow run our automated benchmarking suite and it shows a 2% improvement.

File: .cicd/platforms/pinned/amazon_linux-2-pinned.dockerfile [hunk 1]
```diff
@@ -14,26 +14,18 @@ RUN curl -LO https://github.com/Kitware/CMake/releases/download/v3.16.2/cmake-3.
...
-    cd ../ && git clone --single-branch --branch release_80 https://git.llvm.org/git/libunwind.git && cd libunwind && git checkout 57f6739 && \
-    cd ../ && git clone --single-branch --branch release_80 https://git.llvm.org/git/compiler-rt.git && cd compiler-rt && git checkout 5bc7979 && \
-    mkdir /clang8/build && cd /clang8/build && \
-    cmake -G 'Unix Makefiles' -DCMAKE_INSTALL_PREFIX='/usr/local' -DLLVM_BUILD_EXTERNAL_COMPILER_RT=ON -DLLVM_BUILD_LLVM_DYLIB=ON -DLLVM_ENABLE_LIBCXX=ON -DLLVM_ENABLE_RTTI=ON -DLLVM_INCLUDE_DOCS=OFF -DLLVM_OPTIMIZED_TABLEGEN=ON -DLLVM_TARGETS_TO_BUILD=X86 -DCMAKE_BUILD_TYPE=Release .. && \
+# build clang10
+RUN git clone --single-branch --branch llvmorg-10.0.0 https://github.com/llvm/llvm-project clang10 && \
+    mkdir /clang10/build && cd /clang10/build && \
+    cmake -G 'Unix Makefiles' -DCMAKE_INSTALL_PREFIX='/usr/local' -DLLVM_ENABLE_PROJECTS='lld;polly;clang;clang-tools-extra;libcxx;libcxxabi;libunwind;compiler-rt' -DLLVM_BUILD_LLVM_DYLIB=ON -DLLVM_ENABLE_RTTI=ON -DLLVM_INCLUDE_DOCS=OFF -DLLVM_TARGETS_TO_BUILD=host -DCMAKE_BUILD_TYPE=Release ../llvm && \
```
Comment (johndebord) [inline .cicd/platforms/pinned/amazon_linux-2-pinned.dockerfile:20]: Do we need to build `clang-tools-extra`? Would speed up the build just a tad if we don't.

Comment (spoonincode) [reply .cicd/platforms/pinned/amazon_linux-2-pinned.dockerfile:20]: I can try w/o it and see if it works. Originally I was just doing a 1:1 with what was already there.

Comment (johndebord) [reply .cicd/platforms/pinned/amazon_linux-2-pinned.dockerfile:20]: I'm not sure what we would need it for. But if it sped up the build time, it'd still be a small win. But judging off of the tests so far, it looks like building on Mac is having issues?

Comment (spoonincode) [reply .cicd/platforms/pinned/amazon_linux-2-pinned.dockerfile:20]: All the builds were okay (mac issue was something fleet related). So we don't _need_ it... but on the other hand I think the idea was this was a full featured toolchain. Probably okay to just keep it in for this PR as that's how it was previously. We can remove it in the future (even soon).

**Pull request merged**